vyper/ast/folding.py

```diff
@@ -172,20 +172,39 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
             # annotation is not wrapped in `constant(...)`
             continue
 
-        changed_nodes += replace_constant(vyper_module, node.target.id, node.value, False)
+        # Extract typedef for propagation to type checker
+        node_annotation_arg = node.get("annotation.args")[0]
+        type = None
+        if node_annotation_arg.ast_type == "Subscript":
+            type = node_annotation_arg.value.id
+            # If DynArray, extract typedef from elements
+            if type == "DynArray":
+                type = node_annotation_arg.slice.value.elements[0].id
+        elif node_annotation_arg.ast_type == "Name":
+            type = node_annotation_arg.id
+
+        changed_nodes += replace_constant(
+            vyper_module, node.target.id, node.value, False, type=type
+        )
 
     return changed_nodes
 
 
 # TODO constant folding on log events
 
 
-def _replace(old_node, new_node):
+def _replace(old_node, new_node, type=None):
     if isinstance(new_node, vy_ast.Constant):
-        return new_node.from_node(old_node, value=new_node.value)
+        new_node = new_node.from_node(old_node, value=new_node.value)
+        if type:
+            new_node._metadata["type"] = type
```

Comment:
isn't `node._metadata["type"]` usually some kind of type object, not a string?

Reply:
Yes, it is usually a `TypeDefinition`. My initial thinking was that `TypeDefinition` objects are assigned at a later stage under `semantics`, and it could be inappropriate/messy/premature to import those classes and helper functions into the `ast` module. However, as you pointed out, the inconsistency/messiness ends up in the `_metadata["type"]` instead.

Reply:
Right that's very interesting, I forgot type checking happens after constant folding. We should consider switching the order of the phases.

Reply:
I tried moving the type check phase before constant folding. The good news is that it resolves the issue being addressed in this PR. The bad news is that there are about another 100 test cases that fail. 

Some of the type-checking issues raised are due to:
- Values for built-in constants such as `ZERO_ADDRESS` not being injected, and therefore not being 
- Values for user-defined constants in cases where they are used as subscript or loop ranges not being injected

The remaining issues will require more investigation (e.g. type mismatches, method id must be folded, missing "type" key in metadata).

Reply:
In that case, we should not use the `"type"` key because having `"type"` be polymorphic is really messy. We should think about a more long-term solution but for now I think we can attach a `"constant_annotation"`, which we can inspect during the annotation / type checking phase. Also I would just attach the entire annotation (AST node) instead of converting it to a string.